sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -0,0 +1,351 @@
+      // Shift `result` to the right, discarding high bits.
+      block_shiftr(result, 8);
+      // Look up the product of `overflow` and the low terms of the modulus in
+      // the precomputed table.
+      uint16_t reduce_term = kGFReduceTable[overflow];
+      // Add (xor) this product to the low bits to complete modular reduction.
+      // This works because (low + x^128 * high) is equivalent to (low + (x^128
+      // - modulus) * high).
+      result->data[0] ^= reduce_term;
+    }
+
+    // Add the product of the next byte and H to `result`. We process the bytes
+    // starting with the most significant polynomial terms, which means
+    // starting from the last byte and proceeding to the first.
+    uint8_t pi = block_byte_get(p, kAesBlockNumBytes - 1 - i);
+    block_xor(result, &H_table[pi], result);
+  }
+}
+
+/**
+ * Advances the GHASH state for the given input.
+ *
+ * This function is convenient for computing the GHASH of large, concatenated
+ * buffers. Given a state representing GHASH(A) and a new input B, this
+ * function will return a state representing GHASH(A || B || <padding for B>).
+ *
+ * Pads the input with 0s on the right-hand side if needed so that the input
+ * size is a multiple of the block size.
+ *
+ * The product table should match the format from `make_product_table`.
+ *
+ * @param H_table Precomputed product table
+ * @param input_len Length of input in bytes
+ * @param input Input buffer
+ * @param state GHASH state, updated in place
+ */
+static void aes_gcm_ghash_update(const aes_block_t H_table[256],
+                                 const size_t input_len, const uint8_t *input,
+                                 aes_block_t *state) {
+  // Calculate the number of AES blocks needed for the input.
+  size_t nblocks = ceiling_div(input_len, kAesBlockNumBytes);
+
+  // Main loop.
+  for (size_t i = 0; i < nblocks; i++) {
+    // Construct block i of the input.
+    aes_block_t input_block;
+    if ((i == nblocks - 1) && (input_len % kAesBlockNumBytes != 0)) {
+      // Last block is not full; pad with zeroes.
+      size_t nbytes = input_len % kAesBlockNumBytes;
+      uint8_t *input_data = (uint8_t *)input_block.data;
+      memcpy(input_data, input, nbytes);
+      memset(input_data + nbytes, 0, kAesBlockNumBytes - nbytes);
+    } else {
+      // Full block; copy data and advance input pointer.
+      memcpy(input_block.data, input, kAesBlockNumBytes);
+      input += kAesBlockNumBytes;
+    }
+    // XOR `state` with the next input block.
+    block_xor(state, &input_block, state);
+    // Multiply state by H and update.
+    aes_block_t yH;
+    galois_mul(state, H_table, &yH);
+    memcpy(state->data, yH.data, kAesBlockNumBytes);
+  }
+}
+
+aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
+                          const size_t input_len, const uint8_t *input,
+                          aes_block_t *output) {
+  // Compute the product table for H.
+  aes_block_t H_table[256];
```

Comment:
For an n-bit window, this method requires:
- A reduction table (constant) with `2^n` elements, `8+n` bits each
- A product table (computed once per encryption or decryption operation) with `2^n` elements, 128 bits each

So the total size for this implementation with an 8-bit window is 4.6 kiB. It can pretty straightforwardly be adapted to use a 4-bit window instead, which would be much smaller (about 280 bytes) but slower. Let me actually try it out and benchmark it so we know what the tradeoffs are; I think it will be significantly slower but I'm not certain how much. We could also consider setting things up so that one can choose which setup to use at compile time -- space vs speed.

Reply:
Okay, here's some rough benchmarks:

| Window size  | Storage (bytes) | Overhead (cycles) | GHASH (cycles) |
| --------- | -------- | ------------- |-------------|
|  4 bits  |   280  | 17,100 | 20,100 |
|  8 bits  |   4600 |  170,300 | 10,700 |

The "overhead" happens once per encryption (and is dominated in the 8-bit case by computing the product table). The GHASH happens 1-2x per encryption (depending on IV length) and then 1x per 128-bit block of plaintext OR aad. So the 4-bit window is faster for smaller messages, but scales much worse:

|  Plaintext + aad size | 4 bits (cycles / encrypt) | 8 bits (cycles / encrypt) |
| ------------- | ----------- | ----------- |
|  0 bytes    |   57,200 | 191,700 |
| 128 bytes | 218,000 | 277,300 |
| 256 bytes | 378,800 | 362,900 |
| 384 bytes | 539,600 | 448,500 |
| 512 bytes | 700,400 | 534,100 |
...
...
...
| 1024 bytes | 1,343,600 | 876,500 |

However, looking at these cycle counts, we're already looking at quite high latency for these input sizes; at 100 MHz, the 512-byte case takes 5 ms even for the 8-bit version. Since we're not offering a streaming API, this is probably pushing the edge of where latency requirements would require chunking the input anyway.

An important note here is that I believe using bitmanip instructions will remove the need for precomputed tables of any size.

Bottom line: Given the relatively late point at which speed benefits appear, it's probably worth prioritizing space and switching to a 4-bit window. Long-term, bitmanip instructions will replace tables.

Reply:
I commented on the same line :) We may also consider moving this to `.bss`, i.e. making this a `static`.

Reply:
SGTM. I agree using 4bit is a good tradeoff for a memory versus performance perspective.

As a side note, I assume that if we end up implementing hardening on this, we'll choose the implementation that uses bitmanip instructions.